src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -88,6 +88,7 @@ CircularBoundedJournal::do_submit_record(
   if (action == RecordSubmitter::action_t::ROLL) {
     return record_submitter.roll_segment(
     ).safe_then([this, record=std::move(record), &handle]() mutable {
+      record_submitter.update_committed_to(get_written_to());
```

Comment:
I don't understand how this is correct.

Please refer to another comment, I think we should update committed_to only after a record is written, which is already implemented below. IIRC having rolled a segment is not related to whether a previous record is fully written.

Reply:
If CBJ writes a record after rolled, the record's committed_to is the end of the journal, as below.
<----------------CBJ-------------------->
|--new record--|--------------------------|
^start                                                              ^ end


After that, at replay() time, we need to replay the record which is located at the start address, but
it's committed_to indicates the end (also the seq value is cursor's current seq -1). So, there is no way to identify the record is either the first record in the journal or the written record right after the roll occurs. 
How can we identify the rolled record without this change?

Reply:
> So, there is no way to identify the record is either the first record in the journal or the written record right after the roll occurs.

Replay starts at journal tail. Is it reasonable to assume that its right side should contain records before rolling, and its left part might contain rolled records with the increased `segment_seq`?

In another word, it looks reasonable to assume that the first record is not rolled if and only if the journal tail is at 0.

Reply:
Yes, but the first record's committed_to, which was recorded right after rolled, is the end address of the device (segment_seq is not increased in this case) if the tail is in the middle of journal space and the roll happened before.


So, how about modifying read_validate_record_metadata, as follows:

````
RecordScanner::read_validate_record_metadata_ret CircularBoundedJournal::read_validate_record_metadata(
  scan_valid_records_cursor &cursor,
  segment_nonce_t nonce)
{
  LOG_PREFIX(CircularBoundedJournal::read_validate_record_metadata);
  paddr_t start = cursor.seq.offset;
  return read_record(start, nonce
  ).safe_then([FNAME, &cursor, this](auto ret) {
    if (!ret.has_value()) {
      return read_validate_record_metadata_ret(
        read_validate_record_metadata_ertr::ready_future_marker{},
        std::nullopt);
    }
    auto [r_header, bl] = *ret;
    auto tail = get_dirty_tail() <= get_alloc_tail() ?
      get_dirty_tail() : get_alloc_tail();
    journal_seq_t expected_committed;
    if (cursor.last_committed != JOURNAL_SEQ_NULL) {
      expected_committed.segment_seq = cursor.last_committed.segment_seq;
      expected_committed.offset =
        cursor.last_committed.offset.add_offset(r_header.mdlength + r_header.dlength);
    }
    if (r_header.committed_to.segment_seq != cursor.seq.segment_seq &&
        (convert_paddr_to_abs_addr(tail.offset) == get_records_start() ||
        expected_committed != r_header.committed_to)) {
      DEBUG("invalid header: {}", r_header);
      return read_validate_record_metadata_ret(
        read_validate_record_metadata_ertr::ready_future_marker{},
        std::nullopt);
    }

    bufferlist mdbuf;
    mdbuf.substr_of(bl, 0, r_header.mdlength);
    DEBUG("header: {}", r_header);
    return read_validate_record_metadata_ret(
      read_validate_record_metadata_ertr::ready_future_marker{},
      std::make_pair(std::move(r_header), std::move(mdbuf)));
  });
}
```